Fix data race in volume controller unit test.

jsafrane opened the pull request:
Reactor must be locked when fiddling with reactor.volumes and reactor.claims. Therefore add new functions to add/delete volume/claim with sending an event.

Fixes #26345

jsafrane commented:
Raising priority to get merged sooner, everybody likes fixed flakes, right?

jsafrane commented:
And I wonder, what does ' e2e-not-required` label means, does it skip **only** e2e tests? I'm fixing unit test only.

saad-ali commented:
Assigning to @jingxu97 for review.

jingxu97 commented on pkg/controller/persistentvolume/framework_test.go:
```diff
@@ -471,6 +471,62 @@ func (r *volumeReactor) waitTest() {
+
+	// Generate deletion event. Cloned volume is needed to prevent races (and we
+	// would get a clone from etcd too).
+	clone, _ := conversion.NewCloner().DeepCopy(claim)
```
I am wondering whether the clone operation in this func and deleteVolumeWithEvent function can be put before the lock?

jsafrane replied:
It could, however... would the code be more effective? I doubt so. Would it be more readable? IMO it's clearer as it is.

jingxu97 commented on pkg/controller/persistentvolume/framework_test.go:
```diff
@@ -471,6 +471,62 @@ func (r *volumeReactor) waitTest() {
 
+// deleteVolumeWithEvent simulates that a volume has been deleted in etcd and
+// the controller receives 'volume deleted' event.
+func (r *volumeReactor) deleteVolumeWithEvent(volume *api.PersistentVolume) {
```
Could we remove WithEvent from the function name unless it may cause confusion?

jsafrane replied:
These functions do not really add/delete volumes/claims, they generate fake events. , `addClaim` would be confusing. I'll rename them to `addClaimEvent`.

jsafrane committed:
Fix data race in volume controller unit test.

Reactor must be locked when fiddling with reactor.volumes and reactor.claims.
Therefore add new functions to add/delete volume/claim with sending an event.

jsafrane commented:
rebased and renamed functions

jingxu97 commented on pkg/controller/persistentvolume/controller_test.go:
```diff
@@ -92,14 +83,7 @@ func TestControllerSync(t *testing.T) {
-				clone, _ := conversion.NewCloner().DeepCopy(volume)
-				volumeClone := clone.(*api.PersistentVolume)
-				reactor.volumeSource.Delete(volumeClone)
+				reactor.deleteVolumeEvent(volume)
```
Just want to make sure, we don't need add any lock on line 103 and 107 (in your new version), right?

jsafrane replied:
no, we do not need any lock there.

jingxu97 commented:
LGTM
